File: engine/overlap.py
"""Detección de traslapes entre sesiones (día + rango horario)."""
from __future__ import annotations


def sesiones_se_traslapan(a: list, b: list) -> bool:
    """True si alguna sesión de `a` se traslapa en día+horario con alguna de `b`.

    Dos sesiones que solo se tocan (una termina 10:30 y la otra empieza 10:30)
    NO se consideran traslape: la comparación es estricta.
    """
    for s1 in a:
        for s2 in b:
            if s1.dias & s2.dias:
                if s1.inicio_min < s2.fin_min and s2.inicio_min < s1.fin_min:
                    return True
    return False

File: web/src/lib/engine/overlap.ts
/** Detección de traslapes entre sesiones (día + rango horario).
 * Puerto 1:1 de engine/overlap.py. */
import type { Sesion } from "./models";

/** True si alguna sesión de `a` se traslapa en día+horario con alguna de `b`.
 * Dos sesiones que solo se tocan (una termina 10:30 y la otra empieza 10:30)
 * NO se consideran traslape: la comparación es estricta. */
export function sesionesSeTraslapan(a: Sesion[], b: Sesion[]): boolean {
  for (const s1 of a) {
    for (const s2 of b) {
      if (s1.dias.some((d) => s2.dias.includes(d))) {
        if (s1.inicio_min < s2.fin_min && s2.inicio_min < s1.fin_min) return true;
      }
    }
  }
  return false;
}
